From abbotti@mev.co.uk  Wed Jan 27 14:27:00 2010
From: Ian Abbott <abbotti@mev.co.uk>
Date: Wed, 20 Jan 2010 13:04:44 +0000
Subject: Staging: comedi: pcl812: Correct AI scan counting
Cc: Frank Mori Hess <fmhess@users.sourceforge.net>, Ian Abbott <abbotti@mev.co.uk>, Greg Kroah-Hartman <gregkh@suse.de>
Message-ID: <1263992691-8950-2-git-send-email-abbotti@mev.co.uk>


The AI scan counter should be updated after every completed scan,
not after every channel.  Keep track of current channel.

Signed-off-by: Ian Abbott <abbotti@mev.co.uk>
Signed-off-by: Greg Kroah-Hartman <gregkh@suse.de>

---
 drivers/staging/comedi/drivers/pcl812.c |    8 ++++++--
 1 file changed, 6 insertions(+), 2 deletions(-)

--- a/drivers/staging/comedi/drivers/pcl812.c
+++ b/drivers/staging/comedi/drivers/pcl812.c
@@ -995,7 +995,9 @@ static irqreturn_t interrupt_pcl812_ai_i
 
 	outb(0, dev->iobase + PCL812_CLRINT);	/* clear INT request */
 
-	if (s->async->cur_chan == 0) {	/* one scan done */
+	s->async->cur_chan++;
+	if (s->async->cur_chan >= devpriv->ai_n_chan) {	/* one scan done */
+		s->async->cur_chan = 0;
 		devpriv->ai_act_scan++;
 		if (!(devpriv->ai_neverending))
 			if (devpriv->ai_act_scan >= devpriv->ai_scans) {	/* all data sampled */
@@ -1021,7 +1023,9 @@ static void transfer_from_dma_buf(struct
 	for (i = len; i; i--) {
 		comedi_buf_put(s->async, ptr[bufptr++]);	/*  get one sample */
 
-		if (s->async->cur_chan == 0) {
+		s->async->cur_chan++;
+		if (s->async->cur_chan >= devpriv->ai_n_chan) {
+			s->async->cur_chan = 0;
 			devpriv->ai_act_scan++;
 			if (!devpriv->ai_neverending)
 				if (devpriv->ai_act_scan >= devpriv->ai_scans) {	/* all data sampled */
